File: 프로그래머스/unrated/133502. 햄버거 만들기/햄버거 만들기.cpp
#include <vector>
#include <stack>

using namespace std;

int solution(vector<int> ingredient) {
    int cnt = 0;  // 포장한 햄버거의 개수
    stack<int> st;  // 재료를 저장할 스택

    for (int i = 0; i < ingredient.size(); i++) {
        st.push(ingredient[i]);
        if (st.size() >= 4 && st.top() == 1) {
            int top1 = st.top(); st.pop();
            int top2 = st.top(); st.pop();
            int top3 = st.top(); st.pop();
            int top4 = st.top(); st.pop();
            if (top2 == 3 && top3 == 2 && top4 == 1) {
                cnt++;
            }
            else{st.push(top4); st.push(top3); st.push(top2); st.push(top1);}
        }
    }

    return cnt;
}

File: 프로그래머스/unrated/133502. 햄버거 만들기/햄버거 만들기.java
import java.util.*;
class Solution {
    public int solution(int[] ingredient) {
     int cnt = 0;  // 포장한 햄버거의 개수
        Stack<Integer> st = new Stack<>();  // 재료를 저장할 스택

        for (int i = 0; i < ingredient.length; i++) {
            st.push(ingredient[i]);
            if (st.size() >= 4
                    && st.peek() == 1
                    && st.get(st.size() - 2) == 3
                    && st.get(st.size() - 3) == 2
                    && st.get(st.size() - 4) == 1
            ) {
                st.pop();   st.pop();   st.pop();  st.pop(); cnt++;
            }
        }

        return cnt;
    }
}
